lib/vrl/compiler/src/expression/function_call.rs

```diff
@@ -167,16 +167,244 @@ impl FunctionCall {
+                                // the ind of the target of the closure.
+                                VariableKind::Target => {
+                                    (target.type_def((local, external)), target.as_value())
+                                }
+
+                                // The variable kind is expected to be equal to
+                                // the recuded kind of all values within the
+                                // target collection type.
+                                //
+                                // This assumes the target is a collection type,
+                                // or else it'll return "any".
+                                VariableKind::TargetInnerValue => {
+                                    let kind = if let Some(object) = type_def.as_object() {
+                                        object.reduced_kind()
+                                    } else if let Some(array) = type_def.as_array() {
+                                        array.reduced_kind()
+                                    } else {
+                                        Kind::any()
+                                    };
+
+                                    (kind.into(), None)
+                                }
+
+                                // The variable kind is expected to be equal to
+                                // the kind of all keys within the target
+                                // collection type.
+                                //
+                                // This means it's either a string for an
+                                // object, integer for an array, or
+                                // a combination of the two if the target isn't
+                                // known to be exactly one of the two.
+                                //
+                                // If the target can resolve to a non-collection
+                                // type, this again returns "any".
+                                VariableKind::TargetInnerKey => {
+                                    let mut kind = Kind::empty();
+
+                                    if !type_def.is_collection() {
+                                        kind = Kind::any()
+                                    } else {
+                                        if type_def.is_object() {
+                                            kind.add_bytes();
+                                        }
+                                        if type_def.is_array() {
+                                            kind.add_integer();
+                                        }
+                                    }
+
+                                    (kind.into(), None)
+                                }
+                            };
+
+                            let details = Details { type_def, value };
+
+                            local.insert_variable(call_ident.to_owned().into_inner(), details);
```

Comment:
I'm not seeing anywhere where this variable is cleaned up after the function call, so unless I'm mistaken the typedef for the closure parameters will persist after function call. This needs to be reworked inline with the scoping definitions set out in the rfc.

This is also going to be a problem for any variables defined in the closure block. Since their values should not persist outside the closure, neither should their definitions.

Reply:
Resolved in https://github.com/vectordotdev/vector/pull/12384.